iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/service/TopologyService.java

```diff
@@ -0,0 +1,198 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class TopologyService {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TopologyService.class);
+  private static final long PROBING_INTERVAL_MS = 5_000L;
+  private static final long PROBING_TIMEOUT_MS = 1_000L;
+  private final ScheduledExecutorService topologyExecutor =
+      IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor(
+          ThreadName.CONFIG_NODE_TOPOLOGY_SERVICE.getName());
+
+  private final Consumer<Map<Integer, Set<Integer>>> topologyChangeListener;
+  private final IManager configManager;
+  private ScheduledFuture<?> future;
+
+  /* (fromDataNodeId, toDataNodeId) -> heartbeat history */
+  private final Map<Pair<Integer, Integer>, List<AbstractHeartbeatSample>> heartbeats;
+
+  private final IFailureDetector failureDetector;
+  private static final ConfigNodeConfig CONF = ConfigNodeDescriptor.getInstance().getConf();
+
+  public TopologyService(
+      IManager configManager, Consumer<Map<Integer, Set<Integer>>> topologyChangeListener) {
+    this.configManager = configManager;
+    this.topologyChangeListener = topologyChangeListener;
+    this.heartbeats = new HashMap<>();
+
+    // here we use the same failure
+    switch (CONF.getFailureDetector()) {
+      case IFailureDetector.PHI_ACCRUAL_DETECTOR:
+        this.failureDetector =
+            new PhiAccrualDetector(
+                CONF.getFailureDetectorPhiThreshold(),
+                CONF.getFailureDetectorPhiAcceptablePauseInMs() * 1000_000L,
+                CONF.getHeartbeatIntervalInMs() * 200_000L,
+                60,
+                new FixedDetector(CONF.getFailureDetectorFixedThresholdInMs() * 1000_000L));
+        break;
+      case IFailureDetector.FIXED_DETECTOR:
+      default:
+        this.failureDetector =
+            new FixedDetector(CONF.getFailureDetectorFixedThresholdInMs() * 1000_000L);
+    }
+  }
+
+  public void startTopologyService() {
+    future =
+        ScheduledExecutorUtil.safelyScheduleWithFixedDelay(
+            topologyExecutor, this::topologyProbing, 0, PROBING_INTERVAL_MS, TimeUnit.MILLISECONDS);
+    LOGGER.info("Topology Probing has started successfully");
+  }
+
+  public void stopTopologyService() {
+    future.cancel(true);
+    future = null;
+    LOGGER.info("Topology Probing has stopped successfully");
+  }
+
+  private void topologyProbing() {
+    // 1. get the latest datanode list
+    final List<TDataNodeLocation> dataNodeLocations = new ArrayList<>();
+    final Set<Integer> dataNodeIds = new HashSet<>();
+    final Map<TEndPoint, Integer> endPoint2IdMap = new HashMap<>();
```

Comment:
How about just adding datanodeid to TestDNConnectionRPC?

Reply:
Added.